File: web/src/app/utils/latencyPressure.ts
export const LATENCY_PRESSURE_BLUE = 'var(--cds-support-info)'
export const LATENCY_PRESSURE_RED = 'var(--cds-support-error)'

export type LatencyPressureTone = 'blue' | 'red'
export type LatencyPressureStatus = 'waiting' | 'offline' | 'stable' | 'watch' | 'critical'

export interface LatencyPressureInputs {
  running?: boolean | null
  totalLatencyMs?: number | null
  rtlP95Ms?: number | null
  jitterP95Ms?: number | null
  xrunCount?: number | null
  callbackBudgetMs?: number | null
  currentCallbackMs?: number | null
  headroomPercent?: number | null
}

export interface LatencyPressureAnalysis {
  isAvailable: boolean
  score: number | null
  scoreDisplay: string
  weightedScore: number | null
  pressurePercent: number | null
  tone: LatencyPressureTone
  toneColor: string
  fillColor: string
  status: LatencyPressureStatus
  statusLabel: string
  inputs: {
    running: boolean | null
    totalLatencyMs: number | null
    rtlP95Ms: number | null
    effectiveLatencyMs: number | null
    jitterP95Ms: number | null
    xrunCount: number | null
    callbackBudgetMs: number | null
    currentCallbackMs: number | null
    callbackRatio: number | null
    callbackLoadPercent: number | null
    headroomPercent: number | null
  }
  components: {
    callback: number
    latency: number
    jitter: number
    headroom: number
    xruns: number
  }
}

function clamp(value: number, min: number, max: number): number {
  if (!Number.isFinite(value)) {
    return min
  }

  return Math.min(max, Math.max(min, value))
}

function toFiniteNumber(value: number | null | undefined): number | null {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    return null
  }

  return value
}

function scoreBadMetric(value: number | null, goodThreshold: number, badThreshold: number): number {
  if (value == null) {
    return 10
  }

  if (value <= goodThreshold) {
    return 10
  }

  if (value >= badThreshold) {
    return 0
  }

  const normalized = (value - goodThreshold) / (badThreshold - goodThreshold)
  return clamp(10 * (1 - normalized), 0, 10)
}

function scoreGoodMetric(value: number | null, badThreshold: number, goodThreshold: number): number {
  if (value == null) {
    return 10
  }

  if (value >= goodThreshold) {
    return 10
  }

  if (value <= badThreshold) {
    return 0
  }

  const normalized = (value - badThreshold) / (goodThreshold - badThreshold)
  return clamp(10 * normalized, 0, 10)
}

function xrunScore(xrunCount: number | null): number {
  if (xrunCount == null || xrunCount <= 0) {
    return 10
  }

  return 0
}

function xrunScoreCap(xrunCount: number | null): number {
  if (xrunCount == null || xrunCount <= 0) {
    return 10
  }

  if (xrunCount === 1) {
    return 6
  }

  if (xrunCount <= 3) {
    return 4
  }

  return 2
}

function describeStatus(score: number | null, running: boolean | null, isAvailable: boolean): {
  status: LatencyPressureStatus
  statusLabel: string
} {
  if (!isAvailable || score == null) {
    return { status: 'waiting', statusLabel: 'Waiting' }
  }

  if (running === false) {
    return { status: 'offline', statusLabel: 'Offline' }
  }

  if (score <= 3) {
    return { status: 'critical', statusLabel: 'Critical' }
  }

  if (score <= 7) {
    return { status: 'watch', statusLabel: 'Watch' }
  }

  return { status: 'stable', statusLabel: 'Stable' }
}

export function formatLatencyPressureScore(score: number | null | undefined): string {
  if (typeof score !== 'number' || !Number.isFinite(score)) {
    return '--'
  }

  return String(clamp(Math.round(score), 0, 10)).padStart(2, '0')
}

export function computeLatencyPressure(inputs: LatencyPressureInputs): LatencyPressureAnalysis {
  const running = typeof inputs.running === 'boolean' ? inputs.running : null
  const totalLatencyMs = toFiniteNumber(inputs.totalLatencyMs)
  const rtlP95Ms = toFiniteNumber(inputs.rtlP95Ms)
  const jitterP95Ms = toFiniteNumber(inputs.jitterP95Ms)
  const xrunCount = toFiniteNumber(inputs.xrunCount)
  const callbackBudgetMs = toFiniteNumber(inputs.callbackBudgetMs)
  const currentCallbackMs = toFiniteNumber(inputs.currentCallbackMs)
  const headroomPercent = toFiniteNumber(inputs.headroomPercent)
  const effectiveLatencyMs = totalLatencyMs != null || rtlP95Ms != null
    ? Math.max(totalLatencyMs ?? 0, rtlP95Ms ?? 0)
    : null
  const callbackRatio = callbackBudgetMs != null && callbackBudgetMs > 0 && currentCallbackMs != null
    ? currentCallbackMs / callbackBudgetMs
    : null
  const callbackLoadPercent = callbackRatio != null ? clamp(Math.round(callbackRatio * 100), 0, 999) : null
  const hasAnyMetric = Boolean(
    effectiveLatencyMs != null ||
    jitterP95Ms != null ||
    callbackRatio != null ||
    headroomPercent != null ||
    (xrunCount != null && xrunCount > 0),
  )

  if (!hasAnyMetric) {
    const waitingStatus = describeStatus(null, running, false)

    return {
      isAvailable: false,
      score: null,
      scoreDisplay: formatLatencyPressureScore(null),
      weightedScore: null,
      pressurePercent: null,
      tone: 'blue',
      toneColor: LATENCY_PRESSURE_BLUE,
      fillColor: 'color-mix(in srgb, var(--cds-support-info) 18%, transparent)',
      status: waitingStatus.status,
      statusLabel: waitingStatus.statusLabel,
      inputs: {
        running,
        totalLatencyMs,
        rtlP95Ms,
        effectiveLatencyMs,
        jitterP95Ms,
        xrunCount,
        callbackBudgetMs,
        currentCallbackMs,
        callbackRatio,
        callbackLoadPercent,
        headroomPercent,
      },
      components: {
        callback: 10,
        latency: 10,
        jitter: 10,
        headroom: 10,
        xruns: 10,
      },
    }
  }

  if (running === false) {
    const offlineStatus = describeStatus(0, false, true)

    return {
      isAvailable: true,
      score: 0,
      scoreDisplay: formatLatencyPressureScore(0),
      weightedScore: 0,
      pressurePercent: 100,
      tone: 'red',
      toneColor: LATENCY_PRESSURE_RED,
      fillColor: 'color-mix(in srgb, var(--cds-support-error) 18%, transparent)',
      status: offlineStatus.status,
      statusLabel: offlineStatus.statusLabel,
      inputs: {
        running,
        totalLatencyMs,
        rtlP95Ms,
        effectiveLatencyMs,
        jitterP95Ms,
        xrunCount,
        callbackBudgetMs,
        currentCallbackMs,
        callbackRatio,
        callbackLoadPercent,
        headroomPercent,
      },
      components: {
        callback: 0,
        latency: 0,
        jitter: 0,
        headroom: 0,
        xruns: 0,
      },
    }
  }

  const componentScores = {
    callback: scoreBadMetric(callbackRatio, 0.45, 1.0),
    latency: scoreBadMetric(effectiveLatencyMs, 4.5, 12.0),
    jitter: scoreBadMetric(jitterP95Ms, 0.12, 0.8),
    headroom: scoreGoodMetric(headroomPercent, 10, 45),
    xruns: xrunScore(xrunCount),
  }

  const weightedScore =
    componentScores.callback * 0.32 +
    componentScores.latency * 0.30 +
    componentScores.jitter * 0.18 +
    componentScores.headroom * 0.10 +
    componentScores.xruns * 0.10

  const cappedScore = Math.min(weightedScore, xrunScoreCap(xrunCount))
  const score = clamp(Math.round(cappedScore), 0, 10)
  const rawPressurePercent = clamp(Math.round((1 - weightedScore / 10) * 100), 0, 100)
  const cappedPressurePercent = clamp(Math.round((1 - score / 10) * 100), 0, 100)
  const pressurePercent = Math.max(rawPressurePercent, cappedPressurePercent)
  const tone: LatencyPressureTone = score <= 3 ? 'red' : 'blue'
  const statusInfo = describeStatus(score, true, true)

  return {
    isAvailable: true,
    score,
    scoreDisplay: formatLatencyPressureScore(score),
    weightedScore: Number(weightedScore.toFixed(2)),
    pressurePercent,
    tone,
    toneColor: tone === 'red' ? LATENCY_PRESSURE_RED : LATENCY_PRESSURE_BLUE,
    fillColor: tone === 'red'
      ? 'color-mix(in srgb, var(--cds-support-error) 18%, transparent)'
      : 'color-mix(in srgb, var(--cds-support-info) 18%, transparent)',
    status: statusInfo.status,
    statusLabel: statusInfo.statusLabel,
    inputs: {
      running,
      totalLatencyMs,
      rtlP95Ms,
      effectiveLatencyMs,
      jitterP95Ms,
      xrunCount,
      callbackBudgetMs,
      currentCallbackMs,
      callbackRatio,
      callbackLoadPercent,
      headroomPercent,
    },
    components: componentScores,
  }
}

File: app/utils/latency_pressure.py
"""
Latency Pressure — first-class, platform-wide realtime-audio health score.

Mirrors web/src/app/utils/latencyPressure.ts verbatim (same thresholds, same
weights, same xrun cap, same status bands) so the score is identical no matter
where it is computed. The frontend uses the TypeScript implementation for
locally-collected telemetry; the backend uses this Python implementation to
publish per-node scores on NodeHealth via the topology endpoint.

Inputs are five realtime metrics scored 0..10 each, then weighted into a single
0..10 operator score. The xrun cap enforces "even one glitch can't be 'stable'".

Outputs:
  score              : 0..10 (rounded), None when no telemetry yet
  pressure_percent   : 100 - score*10 (0 = healthy, 100 = critical)
  status             : waiting | offline | stable | watch | critical

This is purely a math function — no I/O, fully deterministic, safe to call
from any thread.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Optional


LatencyPressureStatus = Literal["waiting", "offline", "stable", "watch", "critical"]


@dataclass(frozen=True)
class LatencyPressureInputs:
    running: Optional[bool] = None
    total_latency_ms: Optional[float] = None
    rtl_p95_ms: Optional[float] = None
    jitter_p95_ms: Optional[float] = None
    xrun_count: Optional[int] = None
    callback_budget_ms: Optional[float] = None
    current_callback_ms: Optional[float] = None
    headroom_percent: Optional[float] = None


@dataclass(frozen=True)
class LatencyPressureAnalysis:
    is_available: bool
    score: Optional[int]
    pressure_percent: Optional[int]
    status: LatencyPressureStatus


def _clamp(value: float, lo: float, hi: float) -> float:
    if value != value:  # NaN
        return lo
    return max(lo, min(hi, value))


def _finite(value: Optional[float]) -> Optional[float]:
    if value is None:
        return None
    try:
        f = float(value)
    except (TypeError, ValueError):
        return None
    if f != f or f in (float("inf"), float("-inf")):
        return None
    return f


def _score_bad_metric(value: Optional[float], good_threshold: float, bad_threshold: float) -> float:
    """Higher input is worse — score 10 at <= good, 0 at >= bad, linear between."""
    if value is None:
        return 10.0
    if value <= good_threshold:
        return 10.0
    if value >= bad_threshold:
        return 0.0
    normalized = (value - good_threshold) / (bad_threshold - good_threshold)
    return _clamp(10.0 * (1.0 - normalized), 0.0, 10.0)


def _score_good_metric(value: Optional[float], bad_threshold: float, good_threshold: float) -> float:
    """Higher input is better — score 0 at <= bad, 10 at >= good, linear between."""
    if value is None:
        return 10.0
    if value >= good_threshold:
        return 10.0
    if value <= bad_threshold:
        return 0.0
    normalized = (value - bad_threshold) / (good_threshold - bad_threshold)
    return _clamp(10.0 * normalized, 0.0, 10.0)


def _xrun_score(xrun_count: Optional[int]) -> float:
    if xrun_count is None or xrun_count <= 0:
        return 10.0
    return 0.0


def _xrun_score_cap(xrun_count: Optional[int]) -> float:
    if xrun_count is None or xrun_count <= 0:
        return 10.0
    if xrun_count == 1:
        return 6.0
    if xrun_count <= 3:
        return 4.0
    return 2.0


def _describe_status(score: Optional[float], running: Optional[bool], is_available: bool) -> LatencyPressureStatus:
    if not is_available or score is None:
        return "waiting"
    if running is False:
        return "offline"
    if score <= 3:
        return "critical"
    if score <= 7:
        return "watch"
    return "stable"


def compute_latency_pressure(inputs: LatencyPressureInputs) -> LatencyPressureAnalysis:
    running = inputs.running if isinstance(inputs.running, bool) else None
    total_latency_ms = _finite(inputs.total_latency_ms)
    rtl_p95_ms = _finite(inputs.rtl_p95_ms)
    jitter_p95_ms = _finite(inputs.jitter_p95_ms)
    xrun_count = _finite(inputs.xrun_count)
    xrun_count_int = int(xrun_count) if xrun_count is not None else None
    callback_budget_ms = _finite(inputs.callback_budget_ms)
    current_callback_ms = _finite(inputs.current_callback_ms)
    headroom_percent = _finite(inputs.headroom_percent)

    if total_latency_ms is not None or rtl_p95_ms is not None:
        effective_latency_ms: Optional[float] = max(total_latency_ms or 0.0, rtl_p95_ms or 0.0)
    else:
        effective_latency_ms = None

    callback_ratio: Optional[float]
    if callback_budget_ms is not None and callback_budget_ms > 0 and current_callback_ms is not None:
        callback_ratio = current_callback_ms / callback_budget_ms
    else:
        callback_ratio = None

    has_any_metric = (
        effective_latency_ms is not None
        or jitter_p95_ms is not None
        or callback_ratio is not None
        or headroom_percent is not None
        or (xrun_count_int is not None and xrun_count_int > 0)
    )

    if not has_any_metric:
        return LatencyPressureAnalysis(
            is_available=False,
            score=None,
            pressure_percent=None,
            status=_describe_status(None, running, False),
        )

    if running is False:
        return LatencyPressureAnalysis(
            is_available=True,
            score=0,
            pressure_percent=100,
            status=_describe_status(0.0, False, True),
        )

    callback_score = _score_bad_metric(callback_ratio, 0.45, 1.0)
    latency_score = _score_bad_metric(effective_latency_ms, 4.5, 12.0)
    jitter_score = _score_bad_metric(jitter_p95_ms, 0.12, 0.8)
    headroom_score = _score_good_metric(headroom_percent, 10.0, 45.0)
    xrun_component = _xrun_score(xrun_count_int)

    weighted_score = (
        callback_score * 0.32
        + latency_score * 0.30
        + jitter_score * 0.18
        + headroom_score * 0.10
        + xrun_component * 0.10
    )

    capped_score = min(weighted_score, _xrun_score_cap(xrun_count_int))
    score = int(_clamp(round(capped_score), 0, 10))
    raw_pressure_percent = int(_clamp(round((1.0 - weighted_score / 10.0) * 100.0), 0, 100))
    capped_pressure_percent = int(_clamp(round((1.0 - score / 10.0) * 100.0), 0, 100))
    pressure_percent = max(raw_pressure_percent, capped_pressure_percent)
    status = _describe_status(float(score), True, True)

    return LatencyPressureAnalysis(
        is_available=True,
        score=score,
        pressure_percent=pressure_percent,
        status=status,
    )


__all__ = [
    "LatencyPressureInputs",
    "LatencyPressureAnalysis",
    "LatencyPressureStatus",
    "compute_latency_pressure",
]
